File: utilslib/src/main/java/com/ashlikun/utils/other/spannable/CustomAlignSpan.java
package com.ashlikun.utils.other.spannable;

import android.graphics.Canvas;
import android.graphics.Paint;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.text.style.ReplacementSpan;

/**
 * 作者　　: 李坤
 * 创建时间: 2018/9/3　16:20
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：居上对齐Span
 */
public class CustomAlignSpan extends ReplacementSpan {

    float alignTopOffset;

    public CustomAlignSpan(float alignTopOffset) {
        this.alignTopOffset = alignTopOffset;
    }

    @Override
    public int getSize(@NonNull Paint paint, CharSequence text, int start, int end, @Nullable Paint.FontMetricsInt fm) {
        text = text.subSequence(start, end);
        return (int) paint.measureText(text.toString());
    }

    @Override
    public void draw(@NonNull Canvas canvas, CharSequence text, int start, int end, float x, int top, int y, int bottom, @NonNull Paint paint) {
        text = text.subSequence(start, end);
        Paint.FontMetricsInt fm = paint.getFontMetricsInt();
        //此处重新计算y坐标，使字体对其
        canvas.drawText(text.toString(), x, y - ((y + fm.descent + y + fm.ascent) - (bottom + top)) + alignTopOffset, paint);
    }

}

File: utilslib/src/main/java/com/ashlikun/utils/other/spannable/CustomAlignSpan.kt
package com.ashlikun.utils.other.spannable

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Paint.FontMetricsInt
import android.text.TextPaint
import android.text.style.ReplacementSpan

/**
 * @author　　: 李坤
 * 创建时间: 2021/12/12 16:33
 * 邮箱　　：496546144@qq.com
 *
 *
 * 功能介绍：居上对齐Span
 */
class CustomAlignSpan(
    var alignTopOffset: Float,
    var foregroundColor: Int?,
    var backgroundColor: Int?
) : ReplacementSpan() {
    override fun getSize(
        paint: Paint,
        text: CharSequence,
        start: Int,
        end: Int,
        fm: FontMetricsInt?
    ): Int {
        var text = text
        text = text.subSequence(start, end)
        return paint.measureText(text.toString()).toInt()
    }

    override fun draw(
        canvas: Canvas,
        text: CharSequence,
        start: Int,
        end: Int,
        x: Float,
        top: Int,
        y: Int,
        bottom: Int,
        paint: Paint
    ) {
        var text = text
        if (foregroundColor != null) {
            paint.color = foregroundColor!!
        }
        if (backgroundColor != null) {
            if (paint is TextPaint) {
                paint.bgColor = backgroundColor!!
            }
        }
        text = text.subSequence(start, end)
        val fm = paint.fontMetricsInt
        //此处重新计算y坐标，使字体对其
        canvas.drawText(
            text.toString(),
            x,
            y - (y + fm.descent + y + fm.ascent - (bottom + top)) + alignTopOffset,
            paint
        )
    }
}